File: frontend/lib/camera/surroundings.ts
import type { DetectedHazard, HazardStreamResult, HazardSurroundings } from "@/lib/camera/hazard-stream";

const NEARBY_THRESHOLD = 0.45;
const STOP_THRESHOLD = 0.9;
const CROWDED_FRONT_THRESHOLD = 5;
const FRONT_POSITIONS = new Set(["ahead", "far ahead"]);

function distanceBand(proximity: number): string {
  if (proximity >= STOP_THRESHOLD) return "very close";
  if (proximity >= NEARBY_THRESHOLD) return "nearby";
  return "distant";
}

function bboxPosition(bbox: DetectedHazard["bbox"]): string {
  const cx = (bbox.x1 + bbox.x2) / 2;
  const cy = (bbox.y1 + bbox.y2) / 2;
  if (cy < 0.22) return "far ahead";
  if (cx < 0.36) return "on the left";
  if (cx > 0.64) return "on the right";
  return "ahead";
}

export function describeSurroundings(hazards: DetectedHazard[]): HazardSurroundings {
  const items = hazards.map((hazard) => ({
    label: hazard.label.toLowerCase(),
    proximity: hazard.proximity,
    distanceBand: hazard.distanceBand ?? distanceBand(hazard.proximity),
    position: hazard.position ?? bboxPosition(hazard.bbox),
    confidence: hazard.confidence,
  }));

  const frontCount = items.filter((item) => FRONT_POSITIONS.has(item.position)).length;
  const crowded = frontCount > CROWDED_FRONT_THRESHOLD;
  const crowdedSummary = crowded ? "Crowded environment — please be careful." : null;

  if (!items.length) {
    return {
      summary: crowdedSummary ?? "Path looks clear — nothing detected ahead.",
      voiceText: null,
      items: [],
      crowded,
      frontCount,
    };
  }

  const grouped = new Map<
    string,
    { label: string; band: string; position: string; count: number; proximity: number }
  >();
  for (const item of items) {
    const key = `${item.label}|${item.distanceBand}|${item.position}`;
    const existing = grouped.get(key);
    if (existing) {
      existing.count += 1;
      existing.proximity = Math.max(existing.proximity, item.proximity);
    } else {
      grouped.set(key, {
        label: item.label,
        band: item.distanceBand,
        position: item.position,
        count: 1,
        proximity: item.proximity,
      });
    }
  }

  const phrases = [...grouped.values()]
    .sort((a, b) => b.proximity - a.proximity)
    .map(({ label, band, position, count }) => {
      const noun = count === 1 ? label : `${count} ${label}s`;
      if (band === "very close") return `${noun} very close ${position}`;
      if (position === "ahead") return `${noun} ${band} ahead`;
      return `${noun} ${band} ${position}`;
    });

  let summary = phrases.map((p) => p.charAt(0).toUpperCase() + p.slice(1)).join(". ") + ".";
  if (crowdedSummary) {
    summary = `${crowdedSummary} ${summary}`;
  } else if (items.some((item) => item.proximity >= STOP_THRESHOLD)) {
    summary = `Please be careful. ${summary}`;
  }

  const needsCareful =
    crowded || items.some((item) => item.proximity >= STOP_THRESHOLD);
  const voiceText = needsCareful ? summary : null;

  return { summary, voiceText, items, crowded, frontCount };
}

export function hazardCarefulVoiceText(
  result: Pick<HazardStreamResult, "voiceText" | "surroundings">
): string | null {
  const text = result.voiceText ?? result.surroundings?.voiceText ?? null;
  if (!text) return null;
  return text.replace(/\bstop\b/gi, "be careful").replace(/\s—\s/g, ". ");
}

File: backend/camera-hazard/surroundings.py
"""Turn YOLO detections into a human-readable surroundings description."""
from __future__ import annotations

from typing import Any

NEARBY_THRESHOLD = 0.45


def distance_band(proximity: float, *, stop_threshold: float) -> str:
    if proximity >= stop_threshold:
        return "very close"
    if proximity >= NEARBY_THRESHOLD:
        return "nearby"
    return "distant"


def bbox_position(bbox: dict[str, float]) -> str:
    cx = (bbox["x1"] + bbox["x2"]) / 2
    cy = (bbox["y1"] + bbox["y2"]) / 2

    if cy < 0.22:
        return "far ahead"
    if cx < 0.36:
        return "on the left"
    if cx > 0.64:
        return "on the right"
    return "ahead"


def describe_surroundings(
    hazards: list[dict[str, Any]],
    *,
    stop_threshold: float,
    crowded_front_threshold: int = 5,
) -> dict[str, Any]:
    front_count = sum(
        1
        for h in hazards
        if (h.get("position") or bbox_position(h.get("bbox") or {})) in ("ahead", "far ahead")
    )
    crowded = front_count > crowded_front_threshold
    crowded_summary = (
        "Crowded environment — please be careful."
        if crowded
        else None
    )

    if not hazards:
        return {
            "summary": crowded_summary or "Path looks clear — nothing detected ahead.",
            "voiceText": None,
            "items": [],
            "crowded": crowded,
            "frontCount": front_count,
        }

    items: list[dict[str, Any]] = []
    grouped: dict[tuple[str, str, str], int] = {}

    for hazard in hazards:
        label = str(hazard.get("label", "object")).lower()
        proximity = float(hazard.get("proximity", 0))
        bbox = hazard.get("bbox") or {}
        band = distance_band(proximity, stop_threshold=stop_threshold)
        position = bbox_position(bbox)
        key = (label, band, position)
        grouped[key] = grouped.get(key, 0) + 1

        items.append(
            {
                "label": label,
                "proximity": proximity,
                "distanceBand": band,
                "position": position,
                "confidence": hazard.get("confidence"),
            }
        )

    phrases: list[str] = []
    ranked_groups = sorted(
        grouped.items(),
        key=lambda item: max(
            float(h.get("proximity", 0))
            for h in hazards
            if str(h.get("label", "")).lower() == item[0][0]
        ),
        reverse=True,
    )
    for (label, band, position), count in ranked_groups:
        noun = label if count == 1 else f"{count} {label}s" if not label.endswith("s") else f"{count} {label}"
        if band == "very close":
            phrases.append(f"{noun} very close {position}")
        elif position == "ahead":
            phrases.append(f"{noun} {band} ahead")
        else:
            phrases.append(f"{noun} {band} {position}")

    summary = ". ".join(p.capitalize() for p in phrases) + "."
    if crowded_summary:
        summary = f"{crowded_summary} {summary}"
    elif any(float(h.get("proximity", 0)) >= stop_threshold for h in hazards):
        summary = f"Please be careful. {summary}"

    needs_careful = crowded or any(
        float(h.get("proximity", 0)) >= stop_threshold for h in hazards
    )
    voice_text = summary if needs_careful else None

    return {
        "summary": summary,
        "voiceText": voice_text,
        "items": items,
        "crowded": crowded,
        "frontCount": front_count,
    }
